packages/core/contracts/oracle/implementation/cross-chain/SinkOracle.sol

```diff
@@ -0,0 +1,35 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity ^0.6.0;
+
+import "./BeaconOracle.sol";
+
+/**
+ * @title Simple implementation of the OracleInterface that is intended to be deployed on non-Mainnet networks and used
+ * to communicate price request data cross-chain with a Source Oracle on Mainnet. An Admin can request prices from
+ * this oracle, which might ultimately request a price from a Mainnet Source Oracle and eventually the DVM.
+ * @dev Admins capable of making price requests to this contract should be OptimisticOracle contracts. This enables
+ * optimistic price resolution on non-Mainnet networks while also providing ultimate security by the DVM on Mainnet.
+ */
+contract SinkOracle is BeaconOracle {
+    constructor(address _finderAddress) public BeaconOracle(_finderAddress) {}
+
+    function requestPrice(
+        bytes32 identifier,
+        uint256 time,
+        bytes memory ancillaryData
+    ) public override {
+        _requestPrice(identifier, time, ancillaryData);
+
+        // TODO: Call Bridge.deposit() to intiate cross-chain price request.
+        // _getBridge().deposit(formattedMetadata);
```

Comment:
Here is where we'd want to create recreate this `javascript` code in Solidity:

```js
const encodedMetaDataProposal = Helpers.abiEncode(
      ["bytes32", "uint256", "bytes"],
      [padRight(identifier, 64), requestTime, ancillaryData]
    );
    const depositData = Helpers.createGenericDepositData(encodedMetaDataProposal);
    const depositTxn = await bridgeSidechain.deposit(chainId, votingResourceSidechainId, depositData, {
      from: depositerAddress
    });
```

Reply:
Which parts are hardcoded? Everything except for the depositData?

Can you point me to Helpers? I don't know exactly what it's doing to format the data.

Last question, what method are you encoding to be called on the other side? requestPrice?

Reply:
So I think we would deploy this contract with the `chainId` and `votingResourceSidechainId` sort of hard coded (although we could probably auto detect the latter):

Look [here at this other PR](https://github.com/UMAprotocol/protocol/pull/2894/files#diff-1f7f92853cda43a48b00c8da51120b1704efe313f79842fb27281a558d74014bR124) for reference

Reply:
So this is how you might do this:
Call abi.encodeWithSignature("YOUR_SIGNATURE_HERE", argument1, argument2, ...) to generate the calldata onchain. Note: this would look like this for requestPrice (not verified):
```
bytes memory callData = abi.encodeWithSignature("requestPrice(bytes32,uint256,bytes)", identifier, timestamp, ancillaryData);
```

Note: to make this a little cheaper, you might want to compute the function selector offchain and just hardcode on-chain.
```
bytes4 selector = bytes4(keccak256(bytes("requestPrice(bytes32,uint256,bytes)"))); // Computed once or even hardcoded
...
bytes memory callData = abi.encodeWithSelector(selector, identifier, timestamp, ancillaryData);
```

Reference: https://docs.soliditylang.org/en/v0.6.12/units-and-global-variables.html#abi-encoding-and-decoding-functions.

Reply:
Ended up not needing to specify the function selector in the contract, only need to pack the params

Reply:
Makes sense!